### ISSUE TITLE
HAProxy 1.8.18/19 thread crashes during reload

### ISSUE THREAD

**Issue by kirankumar1991** [haproxy/haproxy#58]
```
$ haproxy -vv
HA-Proxy version 1.8.19 2019/02/11
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O0 -g -fno-strict-aliasing
  OPTIONS = USE_SLZ=1 USE_OPENSSL=1 USE_STATIC_PCRE=1 USE_PCRE_JIT=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE version : 8.41 2017-07-05
Running on PCRE version : 8.41 2017-07-05
PCRE library supports JIT : yes
Built with multi-threading support.
Encrypted password support via crypt(3): yes
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with OpenSSL version : OpenSSL 1.0.1e-fips 11 Feb 2013
Running on OpenSSL version : OpenSSL 1.0.1e-fips 11 Feb 2013
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[TRACE] trace
	[COMP] compression
	[SPOE] spoe

$ uname -a
Linux [hostname redacted] 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/redhat-release
CentOS Linux release 7.4.1708 (Core)
```

**Snippet of HAProxy config:**
```
global
  log     /var/lib/<redacted>/tmp/log.sock local0 info alert
  stats   socket /var/lib/<redacted>/tmp/run/haproxy.sock mode 600 level admin expose-fd listeners
  stats   timeout 2m
  stats   maxconn 2
  nbproc 1
  nbthread 8

  maxconn 20000
  daemon
  pidfile /var/lib/<redacted>/tmp/run/haproxy.pid
  ssl-dh-param-file /var/lib/<redacted>/keys/dhparam.pem
  tune.ssl.lifetime 86400
  tune.ssl.cachesize 600000
  ssl-default-bind-options no-tls-tickets ssl-min-ver TLSv1.0

```

## Steps to reproduce the behavior
I'm able to consistently reproduce this in one of our Performance environments. However, I've not seen this on my devbox. A noticeable difference is that the Performance environment runs on Centos 7.4 and my devbox runs on Centos 6.8. 

- Reloading of HAProxy multiple times seems to crash a subset of the running threads resulting in core dump files generated. We reload heavily during system start-up and when dependent services restart. Once the system stabilizes, I haven't seen the crash.

## Core dump analysis
```
$ gdb /opt/<redacted>/external/bin/haproxy core.14746
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-100.el7_4.1
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /opt/<redacted>/external/bin/haproxy...done.
[New LWP 14752]
[New LWP 14748]
[New LWP 14746]
[New LWP 14753]
[New LWP 14750]
[New LWP 14751]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/opt/<redacted>/external/bin/haproxy -f /var/lib/<redacted>/runtime/config/ha'.
Program terminated with signal 6, Aborted.
#0  0x00007fcdfe540277 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install <redacted>-0.0.0-1143.x86_64
(gdb) bt
#0  0x00007fcdfe540277 in raise () from /lib64/libc.so.6
#1  0x00007fcdfe541968 in abort () from /lib64/libc.so.6
#2  0x00007fcdfe582d37 in __libc_message () from /lib64/libc.so.6
#3  0x00007fcdfe58b499 in _int_free () from /lib64/libc.so.6
#4  0x00000000004d0609 in deinit_log_buffers () at src/log.c:1387
#5  0x00000000004d04c2 in deinit_log_buffers_per_thread () at src/log.c:1365
#6  0x00000000004cbe94 in run_thread_poll_loop (data=0x20f3a18) at src/haproxy.c:2485
#7  0x00007fcdfefb1e25 in start_thread () from /lib64/libpthread.so.0
#8  0x00007fcdfe608bad in clone () from /lib64/libc.so.6
```

Related code:
```
   1380 /* Deinitialize log buffers used for syslog messages */
   1381 void deinit_log_buffers()
   1382 {
   1383         free(logheader);
   1384         free(logheader_rfc5424);
   1385         free(logline);
   1386         free(logline_rfc5424);
   1387         free(startup_logs);
   1388         logheader         = NULL;
   1389         logheader_rfc5424 = NULL;
   1390         logline           = NULL;
   1391         logline_rfc5424   = NULL;
   1392         startup_logs      = NULL;
   1393 }
   1394
```

**Comment by TimWolla** [comment]
@kirankumar1991 I believe this issue is already fixed in the current development version (incidentally just 10 hours ago): http://git.haproxy.org/?p=haproxy.git;a=commitdiff;h=7c49711d6041d1afc42d5b310ddfd7d6f6817c3c

**Comment by kirankumar1991** [comment]
@TimWolla: Thanks for the quick response. I'll wait for the fix to be back-ported to 1.8 branch.
